File: benchmarks/suite/map_stress/bench.py
"""OBO Benchmark: Map Stress -- Hash map / dictionary heavy workload
Tests: map creation, insertion, lookup, iteration, deletion
This is Python's home turf (dicts are highly optimized in CPython)"""


def main():
    n = 50000

    # Phase 1: Build a large dict
    m = {}
    for i in range(n):
        key = "key_" + str(i)
        m[key] = i * 7 + 3
    print("Phase 1: Built map with " + str(len(m)) + " entries")

    # Phase 2: Lookup every key and accumulate
    s = 0
    for i in range(n):
        key = "key_" + str(i)
        s += m[key]
    print("Phase 2: Sum of all values = " + str(s))

    # Phase 3: Build frequency map from list
    items = []
    for i in range(n):
        items.append(i % 100)

    freq = {}
    for item in items:
        key = str(item)
        if key in freq:
            freq[key] += 1
        else:
            freq[key] = 1
    print("Phase 3: Frequency map has " + str(len(freq)) + " unique keys")

    # Phase 4: Nested map operations
    total = 0
    for rnd in range(200):
        inner = {}
        for j in range(500):
            inner["k" + str(j)] = rnd * 500 + j
        for j in range(500):
            total += inner["k" + str(j)]
    print("Phase 4: Nested map total = " + str(total))

    print("Map stress complete.")


main()

File: benchmarks/suite/map_stress/bench.js
function main() {
    var n = 50000;

    // Phase 1: Build a large map
    var m = new Map();
    for (var i = 0; i < n; i++) {
        var key = "key_" + i;
        m.set(key, i * 7 + 3);
    }
    console.log("Phase 1: Built map with " + m.size + " entries");

    // Phase 2: Lookup every key and accumulate
    var s = 0;
    for (var i = 0; i < n; i++) {
        var key = "key_" + i;
        s += m.get(key);
    }
    console.log("Phase 2: Sum of all values = " + s);

    // Phase 3: Build frequency map from list
    var items = [];
    for (var i = 0; i < n; i++) {
        items.push(i % 100);
    }

    var freq = new Map();
    for (var idx = 0; idx < items.length; idx++) {
        var key = "" + items[idx];
        if (freq.has(key)) {
            freq.set(key, freq.get(key) + 1);
        } else {
            freq.set(key, 1);
        }
    }
    console.log("Phase 3: Frequency map has " + freq.size + " unique keys");

    // Phase 4: Nested map operations
    var total = 0;
    for (var rnd = 0; rnd < 200; rnd++) {
        var inner = new Map();
        for (var j = 0; j < 500; j++) {
            inner.set("k" + j, rnd * 500 + j);
        }
        for (var j = 0; j < 500; j++) {
            total += inner.get("k" + j);
        }
    }
    console.log("Phase 4: Nested map total = " + total);

    console.log("Map stress complete.");
}

main();
